dashboard/api_edx_cache.py

```diff
@@ -16,42 +16,31 @@
-
-    enrollments = None
-    certificates = None
-    current_grades = None
-    raw_enrollments = None
-    raw_certificates = None
-    raw_current_grades = None
+    """Represents all edX data related to a User"""
+    edx_data_tuple = namedtuple('EdxDataTuple', ['queryset', 'edx_object'])
 
-    def __init__(self, user, program=None, include_raw_data=False):
-        """
-        Fetches the given User's edx data and sets object properties
-
-        Args:
-            user (User): a User object
-            program (Program): an optional Program to filter on
-            include_raw_data (bool): Set to True if raw edx data is needed (in addition to actual edX objects)
-        """
+    def __init__(self, user, program=None):
+        """Instantiates object and fetches all of a user's cached edX data"""
         self.user = user
         self.program = program
-        self.fetch_and_set_edx_data()
-        if include_raw_data:
-            self.fetch_and_set_raw_data()
-
-    def fetch_and_set_edx_data(self):
-        """Fetches edx data and sets object properties"""
-        self.enrollments = models.CachedEnrollment.get_edx_data(self.user, program=self.program)
-        self.certificates = models.CachedCertificate.get_edx_data(self.user, program=self.program)
-        self.current_grades = models.CachedCurrentGrade.get_edx_data(self.user, program=self.program)
-
-    def fetch_and_set_raw_data(self):
-        """Fetches raw cached data and sets object properties"""
-        self.raw_enrollments = list(models.CachedEnrollment.data_qset(self.user, program=self.program))
-        self.raw_certificates = list(models.CachedCertificate.data_qset(self.user, program=self.program))
-        self.raw_current_grades = list(models.CachedCurrentGrade.data_qset(self.user, program=self.program))
+        self.enrollments = self.fetch_and_set_edx_data_tuple(models.CachedEnrollment)
+        self.certificates = self.fetch_and_set_edx_data_tuple(models.CachedCertificate)
+        self.current_grades = self.fetch_and_set_edx_data_tuple(models.CachedCurrentGrade)
+
+    def fetch_and_set_edx_data_tuple(self, model_cls):
+        """Fetches a user's edX data, deserializes it to an edX API object, and returns a named tuple"""
+        queryset = model_cls.data_qset(self.user, program=self.program)
+        edx_object = model_cls.deserialize_edx_data(queryset)
+        return self.edx_data_tuple(queryset, edx_object)
+
+
+class CachedEdxUserSearchData(CachedEdxUserData):
+    """Represents all edX data related to a User for use in search results"""
+    def fetch_and_set_edx_data_tuple(self, model_cls):
+        """Fetches a user's edX data, deserializes it to an edX API object, and returns a named tuple"""
+        queryset = model_cls.user_course_qset(self.user, program=self.program)
+        edx_object = model_cls.deserialize_edx_data([obj.data for obj in queryset])
+        return self.edx_data_tuple(queryset, edx_object)
```

Comment:
this example gets rid of the two separate classes:

```python
from dashboard.api_edx_cache import CachedEdxDataApi
class CachedEdxUserData:
    """Represents all edX data related to a User"""
    edx_data_tuple = namedtuple('EdxDataTuple', ['queryset', 'edx_object'])
    # Reverse CACHED_EDX_MODELS to allow a reference from a model class to a string name
    model_name_map = {v: k for k, v in CachedEdxDataApi.CACHED_EDX_MODELS}

    def __init__(self, user, program=None, prefetch=None):
        """Instantiates object and fetches all of a user's cached edX data"""
        self.user = user
        self.program = program
        prefetch_list = prefetch or []
        self.enrollments = self.get_edx_data_tuple(models.CachedEnrollment, prefetch_list)
        self.certificates = self.get_edx_data_tuple(models.CachedCertificate, prefetch_list)
        self.current_grades = self.get_edx_data_tuple(models.CachedCurrentGrade, prefetch_list)

    def get_edx_data_tuple(self, model_cls, prefetch_list):
        if self.model_name_map[model_cls] not in prefetch_list:
            # We only need the raw data
            queryset = model_cls.data_qset(self.user, program=self.program)
            data_iterable = queryset
        else:
            # We need the raw data as well as some prefetched related data
            queryset = model_cls.user_course_qset(self.user, program=self.program)
            data_iterable = [obj.data for obj in queryset]
        edx_object = model_cls.deserialize_edx_data(data_iterable)
        return self.edx_data_tuple(queryset, edx_object)
```

for MMTrack it would just be normal usage: `CachedEdxUserData(user)`
for search serialization: `CachedEdxUserData(user, program=program, prefetch=['enrollments'])`

we can also change this to use the actual model class to indicate that it needs to be prefetched: `CachedEdxUserData(user, program=program, prefetch=[CachedEnrollments])`. i think the above example is clear enough and takes care of our needs. let me know what you think